src/searching/searching.py

```diff
@@ -1,14 +1,35 @@
+
+# ? O(n)
+
 def linear_search(arr, target):
     # Your code here
+    for i in range(len(arr)):
+        if arr[i] == target:
+            return i
+        
 
 
     return -1   # not found
 
 
-# Write an iterative implementation of Binary Search
+# * Write an iterative implementation of Binary Search
+# ? - O(log n)
+
 def binary_search(arr, target):
 
     # Your code here
+    left = 0
+    right = len(arr) - 1
+    while left <= right:
+        middle = (left + right) // 2
+        if arr[middle] == target:
+            return middle
+        elif arr[middle] > target:
+            # search to the right
+            right = middle - 1
+        else:
+            # search to the left
+            left = middle + 1
 
 
     return -1  # not found
```

Comment:
The only note I want to add here is that you can often reduce your cyclic complexity (i.e. how indented things are, which decreases readability) _slightly_ through the use of `continue` and `break` inside loops. This is especially true of for loops.

If we did this to your code, it would have the effect of moving your else block to a more "flat" position, and could also allow you to use two `if` statements rather than 1 long `if/elif/else` block:

```python
def binary_search(arr, target):
    low = 0
    high = len(arr) - 1

    while low <= high:
        mid = (low+high) // 2

        if arr[mid] == target:
            return mid

        if arr[mid] < target:
            low = mid + 1
            continue

        high = mid

    return -1
```

The final thing you could consider is "failing fast". If, for example, the array passed to your function was empty, you could just return immediately like so:

```python
def binary_search(arr, target):
    if len(arr) == 0:
        return -1  # array empty

    low = 0
    high = len(arr) - 1

    while low <= high:
        mid = (low+high) // 2

        if arr[mid] == target:
            return mid

        if arr[mid] < target:
            low = mid + 1
            continue

        high = mid

    return -1
```